### PR TITLE
Bug Fixes for downstream of NonSpecific Search and for Partitions

### PR DESCRIPTION
**Pull request by zrolfs**

### CODE REVIEW

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs [hunk 1]
```diff
@@ -124,7 +129,13 @@ public int NumMods
 
         public int MissedCleavages
         {
-            get { return modPep.MissedCleavages; }
```
Comment (stefanks) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: Why do we need a new constructor?

File: EngineLayer/Proteomics/CompactPeptideWithModifiedMass.cs [hunk 2]
```diff
@@ -27,7 +27,9 @@ public CompactPeptideWithModifiedMass(CompactPeptideBase cp, double Monoisotopic
 
         public void AssignCorrectMass()
```
Comment (stefanks) [inline EngineLayer/Proteomics/CompactPeptideWithModifiedMass.cs]: This method is so unintuitive. Maybe rename it? I have no idea what it does

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs [hunk 3]
```diff
@@ -190,7 +201,7 @@ public PeptideWithSetModifications Localize(int j, double massToLocalize)
             }
 
             vvv.Add(j + 2, new ModificationWithMass(null, null, null, TerminusLocalization.Any, massToLocalize + massOfExistingMod, null, new List<double> { 0 }, new List<double> { massToLocalize + massOfExistingMod }, null));
-            var hm = new PeptideWithSetModifications(modPep, vvv, numFixedMods);
+            var hm = new PeptideWithSetModifications(vvv, numFixedMods, this, OneBasedStartResidueInProtein, OneBasedEndResidueInProtein);
```
Comment (stefanks) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: Why the change?

File: EngineLayer/Psm.cs [hunk 4]
```diff
@@ -267,15 +267,25 @@ public void AddOrReplace(CompactPeptideBase compactPeptide, double score, int no
...
-
-                var candidatePli = new ProteinLinkedInfo(matching[ok]);
+                compactPeptides[cpKey] = new Tuple<int, HashSet<PeptideWithSetModifications>>(compactPeptides[cpKey].Item1, matching[cpKey]);
+                var candidatePli = new ProteinLinkedInfo(matching[cpKey]);
+                if (MostProbableProteinInfo == null || FirstIsPreferable(candidatePli, MostProbableProteinInfo))
+                    MostProbableProteinInfo = candidatePli;
+            }
+            List<CompactPeptideBase> duplicateKeysToRemove = new List<CompactPeptideBase>();
+            foreach (KeyValuePair<CompactPeptideBase, Tuple<int, HashSet<PeptideWithSetModifications>>> kvp in compactPeptides)
+                if (kvp.Value.Item2 == null)
```
Comment (stefanks) [inline EngineLayer/Psm.cs]: This should have never been null in the first place. If this condition is true it means there is a bug somewhere before this point.

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs (hunk 3 above)
Comment (zrolfs) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: For NonSpecific search modes, you had previously requested a removal of modPep from newly generated PWSM. When modPep is null, modPep.Protein throws a null pointer exception. The new constructor generates a modPep.Protein by taking this.Protein instead.

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs (hunk 1 above)
Comment (zrolfs) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: When modPep==null, there is a null pointer exception.

File: EngineLayer/Psm.cs (hunk 4 above)
Comment (zrolfs) [inline EngineLayer/Psm.cs]: It's an equal method issue, where there are two identical keys in the dictionary (which should never happen, but occurs when we modify the monoisotopic masses.

File: EngineLayer/Psm.cs [hunk 5]
```diff
@@ -287,6 +286,14 @@ public void MatchToProteinLinkedPeptides(Dictionary<CompactPeptideBase, HashSet<
             Notch = Resolve(compactPeptides.Select(b => b.Value.Item1)).Item2;
         }
 
+        public bool CompactPeptidesContainsKey(CompactPeptideBase key)
+        {
+            if (compactPeptides.Keys.ToList().Contains(key))
```
Comment (stefanks) [inline EngineLayer/Psm.cs]: Inefficient. Just use  compactPeptides.ContainsKey

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs [hunk 6]
```diff
@@ -51,6 +51,11 @@ internal PeptideWithSetModifications(PeptideWithSetModifications modsFromThisOne
             this.allModsOneIsNterminus = modsFromThisOne.allModsOneIsNterminus.Where(b => b.Key >= proteinOneBasedStart - modsFromThisOne.OneBasedStartResidueInProtein && b.Key <= proteinOneBasedEnd - modsFromThisOne.OneBasedEndResidueInProtein).ToDictionary(b => b.Key, b => b.Value);
         }
 
+        internal PeptideWithSetModifications(Dictionary<int, ModificationWithMass> allModsOneIsNterminus, int numFixedMods, PeptideWithSetModifications modsFromThisOne, int proteinOneBasedStart, int proteinOneBasedEnd):base(modsFromThisOne.Protein, proteinOneBasedStart, proteinOneBasedEnd)
```
Comment (stefanks) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: Just pass the Protein object here instead of PeptideWithSetModifications

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs [hunk 7]
```diff
@@ -124,7 +129,13 @@ public int NumMods
 
         public int MissedCleavages
```
Comment (stefanks) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: change to int?

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs [hunk 8]
```diff
@@ -124,7 +129,13 @@ public int NumMods
...
         {
-            get { return modPep.MissedCleavages; }
+            get
+            {
+                if (modPep == null)
+                    return 0;
```
Comment (stefanks) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: return null

**Comment by stefanks** [comment]
Actually, let's just remove the  private readonly PeptideWithPossibleModifications modPep from PWSM! And only populate the int? MissedCleavages property when it is passed in the constructor. This will conserve memory.

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs (hunk 7 above)
Comment (zrolfs) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: it is int?

Comment (stefanks) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: haha
Literally, "int?"

Comment (stefanks) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: nullable

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs (hunk 8 above)
Comment (zrolfs) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: can't. int is nonnullable value type

Comment (acesnik) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: "int?" is nullable :) https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/nullable-types/

Comment (zrolfs) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: aha! so smart.

Comment (stefanks) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: Sorry if I wasn't clear!

File: EngineLayer/Psm.cs [hunk 9]
```diff
@@ -290,6 +289,14 @@ public void MatchToProteinLinkedPeptides(Dictionary<CompactPeptideBase, HashSet<
             Notch = Resolve(compactPeptides.Select(b => b.Value.Item1)).Item2;
         }
 
+        public bool CompactPeptidesContainsKey(CompactPeptideBase key)
+        {
+            if (compactPeptides.ContainsKey(key))
```
Comment (stefanks) [inline EngineLayer/Psm.cs]: just do return compactPeptides.ContainsKey(key)

File: EngineLayer/Proteomics/PeptideWithSetModifications.cs [hunk 10]
```diff
@@ -34,23 +34,30 @@ public class PeptideWithSetModifications : Peptide
...
         {
-            this.modPep = everythingElseFromThisOne.modPep;
+            this.missedCleavages = everythingElseFromThisOne.missedCleavages;
             this.allModsOneIsNterminus = modsFromThisOne.allModsOneIsNterminus;
             this.numFixedMods = modsFromThisOne.numFixedMods;
         }
 
         internal PeptideWithSetModifications(PeptideWithSetModifications modsFromThisOne, int proteinOneBasedStart, int proteinOneBasedEnd) : base(modsFromThisOne.Protein, proteinOneBasedStart, proteinOneBasedEnd)
         {
+            this.missedCleavages = modsFromThisOne.missedCleavages;
```
Comment (stefanks) [inline EngineLayer/Proteomics/PeptideWithSetModifications.cs]: this is incorrect, just leave as null

File: EngineLayer/Psm.cs [hunk 11]
```diff
@@ -315,7 +319,7 @@ public override string ToString()
                 sb.Append('\t' + Resolve(compactPeptides.SelectMany(b => b.Value.Item2).Select(b => b.BaseSequence)).Item1);
                 sb.Append('\t' + Resolve(compactPeptides.SelectMany(b => b.Value.Item2).Select(b => b.Sequence)).Item1);
                 sb.Append('\t' + Resolve(compactPeptides.SelectMany(b => b.Value.Item2).Select(b => b.NumVariableMods)).Item1);
-                sb.Append('\t' + Resolve(compactPeptides.SelectMany(b => b.Value.Item2).Select(b => b.MissedCleavages)).Item1);
+                sb.Append('\t' + Resolve(compactPeptides.SelectMany(b => b.Value.Item2).Select(b => b.missedCleavages ?? -1)).Item1);
```
Comment (stefanks) [inline EngineLayer/Psm.cs]: oh, no need for "?? -1" here!!
